frame/examples/offchain-worker-ping-pong/src/lib.rs

```diff
@@ -0,0 +1,632 @@
+		/// we make sure that some particular calls are being whitelisted and marked as valid.
+		///
+		/// ⚠ WARNING ⚠
+		/// Anyone could be sending these unsigned transactions, not only OCWs!
+		///
+		/// When it comes to signed payloads, **we only check if the signature is coherent with the signer,
+		/// but we don't really check if the signer is an authorized OCW**!
+		///
+		/// You should not interpret signed payloads as a filter that only allows transactions from
+		/// authorized OCWs. Anyone could have signed those payloads, even malicious actors trying
+		/// to "impersonate" an OCW.
+		///
+		/// There are NO implicit security assumptions here!
+		fn validate_unsigned(_source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			// Firstly let's check that we call the right function.
+			if let Call::pong_unsigned_with_signed_payload {
+				pong_payload: ref payload,
+				ref signature,
+			} = call
+			{
+				// ⚠ WARNING ⚠
+				// this is nothing but a "sanity check" on the signature
+				// it only checks if the signature is coherent with the public key of `SignedPayload`
+				// whoever that might be (not necessarily an authorized OCW)
+				let signature_valid =
+					SignedPayload::<T>::verify::<T::AuthorityId>(payload, signature.clone());
+				if !signature_valid {
+					return InvalidTransaction::BadProof.into();
+				}
+				Self::validate_transaction_parameters(&payload.block_number)
+			} else if let Call::pong_unsigned { block_number, nonce: _n } = call {
+				Self::validate_transaction_parameters(block_number)
+			} else {
+				InvalidTransaction::Call.into()
+			}
+		}
+	}
+
+	/// A struct for wrapping the ping nonce.
+	#[derive(Encode, Decode, MaxEncodedLen, TypeInfo)]
+	pub struct Ping(pub u32);
+
+	/// A vector of recently submitted pings.
+	#[pallet::storage]
+	#[pallet::getter(fn pings)]
+	pub(super) type Pings<T: Config> = StorageValue<_, BoundedVec<Ping, T::MaxPings>, ValueQuery>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn authorities)]
+	pub(super) type Authorities<T: Config> =
+		StorageValue<_, BoundedVec<T::AccountId, T::MaxAuthorities>, ValueQuery>;
+
+	/// Defines the block when next unsigned transaction will be accepted.
+	///
+	/// To prevent spam of unsigned (and unpaid!) transactions on the network,
+	/// we only allow one transaction every `T::UnsignedInterval` blocks.
+	/// This storage entry defines when new transaction is going to be accepted.
+	#[pallet::storage]
+	#[pallet::getter(fn next_unsigned_at)]
+	pub(super) type NextUnsignedAt<T: Config> = StorageValue<_, T::BlockNumber, ValueQuery>;
+}
+
+/// Payload used by this example crate to hold pong response
+/// data required to submit an unsigned transaction.
+#[derive(Encode, Decode, Clone, PartialEq, Eq, RuntimeDebug, scale_info::TypeInfo)]
+pub struct PongPayload<Public, BlockNumber> {
+	block_number: BlockNumber,
+	nonce: u32,
+	public: Public,
+}
```

Comment:
I would have this code towards the top of this pallet as it is used within the OCW logic and in the dispatchable calls.